OptaDOS photoemission module

victorchanglee opened the pull request:
Dear all,

This includes a new module for simulating photoemission in OptaDOS.

Details of the implementation are included in documents/photoemission

A .cell .param and .odi example files are also included in examples/Photoemission

Please, let me know about any issue or question about how to use the module.

Kind regards,
Victor

victorchanglee committed:
OptaDOS photoemission module

This pull request includes a new module for simulating photoemission in OptaDOS.

Details of the implementation are included in documents/photoemission

A .cell .param and .odi example files are also included in examples/Photoemission

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+  integer :: max_layer
+  real(kind=dp) :: e_fermi
+  real(kind=dp) :: q_weight 
+  real(kind=dp), parameter :: epsilon_0 = 8.8541878176E-12_dp   !A^2 s^4 kg^-1 m^-3
```
Constants need to live in constants.f90 and be consistent throughout the code, please.

victorchanglee replied:
Thanks, will move that to the appropriate file. It was initially coded so that the original optados code is modified as little as possible

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+  subroutine make_optical_weights    
+    !***************************************************************
+    !This subroutine is copied from the OPTICS module
+
```
I don't understand. Why not call the code? I don't like have copies of the same routines in different places.

victorchanglee replied:
It is a slightly modified version of the optics module. It does a projected or weighted optical properties to actually computed the optical absorption of each atom in order to study the intensity of the light as it travels inside a material

ajm143 replied:
Ok -- makes sense.

ajm143 commented on optados/examples/Si2_PDIS/Si2.bib:
```diff
@@ -0,0 +1,50 @@
+@article{DFT-HK,
```
This shouldn't be checked in.

ajm143 commented on optados/examples/Si2_PDIS/Si2-out.cell:
```diff
@@ -0,0 +1,46 @@
+# CELL file written 10:16:14 (GMT+2.0) 18th July 2020 from run Si2
```
This shouldn't be checked in.

ajm143 commented on optados/examples/Photoemission/Cu.odi:
```diff
@@ -0,0 +1,20 @@
+TASK              : photoemission
```
Might be worth commenting this file more, to explain what the new keywords do here.

ajm143 commented on optados/examples/Photoemission/Cu.odi:
```diff
@@ -0,0 +1,20 @@
+OPTICS_GEOM : unpolar
+optics_qdir : 1 1.000 1.0000
+photon_energy : 4.7
+linear_smearing : 0.026
```
Why do we have both smearing set here?

ajm143 commented on optados/examples/Photoemission/Cu.cell:
```diff
@@ -0,0 +1,29 @@
+
+symmetry_generate
+
+fix_all_cell : true
```
Is this important for the calculation? If not, please can we have a minimal working example ?

ajm143 commented on optados/documents/photoemission/user_guide.tex:
```diff
@@ -0,0 +1,970 @@
+\documentclass[a4paper,11pt,twoside]{book}
```
I'm not convinced that we should have a separate manual. I'm quite happy for an extra chapter in the manual.

ajm143 replied:
Our longer term plan is to have an html / online manual. But for the moment, I think this should just be a chapter of the manual.

ajm143 commented on optados/examples/Photoemission/Cu.param:
```diff
@@ -0,0 +1,17 @@
+SPIN_POLARIZED       : FALSE
+WRITE_FORMATTED_POTENTIAL : TRUE
+WRITE_FORMATTED_DENSITY : TRUE
+MIXING_SCHEME : Pulay
```
Are all of these parameters necessary? It would be helpful not to have red-herrings in here.

victorchanglee replied:
You are right, they are not a requirement. They has to do more with other results in my work.

ajm143 commented on optados/examples/Si2_PDIS/Si2.odo:
```diff
@@ -0,0 +1,93 @@
+ +============================================================================+
+ +                 Projected Dispersion Curve Calculation                     +
+ +============================================================================+
+
```
By all means have a section here that says, "Photoemission by V.C. N.H. et al, please cite..."

ajm143 commented on optados/src/build.f90:
```diff
@@ -0,0 +1,17 @@
+module od_build
```
Please don't check this in.

ajm143 commented on optados/src/electronic.f90:
```diff
@@ -294,6 +304,123 @@ subroutine elec_read_band_gradient
+    return
+
+101 call io_error('Error: Problem opening cst_vel file in read_band_curvature')
+102 call io_error('Error: Problem opening dome_bin file in read_band_curvature')
```
Surely not a dome_bin ?

ajm143 commented:
@victorchanglee I'm confused by a ddome_bin file? What does it contain, and where does it come from? An ome_bin is the optical matrix elements. A dome_bin is the _diagonal_ optical matrix elements, so what does ddome mean? And is that a consistent naming scheme?

ajm143 reviewed (changes requested):
Looks really nice and useful.

These are some initial thoughts on the pull request. I will do more later, but I think it's good for @victorchanglee to have a look. And @Rebecca-Nicholls also needs to see the comments.

Rebecca-Nicholls commented:
@victorchanglee Hi Victor - thank you for all of this!  I'm in the middle of teaching at the moment but should be able to look at it in ~three weeks time.  Andrew has sent you some comments in the meantime.  Best wishes, Rebecca

victorchanglee commented:
> @victorchanglee I'm confused by a ddome_bin file? What does it contain, and where does it come from? An ome_bin is the optical matrix elements. A dome_bin is the _diagonal_ optical matrix elements, so what does ddome mean? And is that a consistent naming scheme?

@ajm143 It contains the second order kp perturbation, which is the band curvature/effective mass. This is currently under "testing"

It is related to how the parallel momentum in ARPES is interpreted. In ARPES, the angle is determined by the parallel momentum of the electrons inside of a crystal. It is usually assign to the crystal momentum or wavevector k. However, the nature of the Bloch theorem makes k only preserved for a defined unit cell or in other words, if we define a supercell, the bands will be folded and thus changing the interpretation of k.

Related to this issue, several band unfolding schemes has being developed (for example, CASTEP has a band unfolding tool), however, in my thesis I took a different look at this momentum interpretation of the band structure.

I used two different approaches, the first one is applying the momentum operator to get the momentum expectation value, which is in fact, the kp first order perturbation already computed in CASTEP in .dome

The second approach I used is a semi-classical momentum, for which I used the band gradient to obtain the electron velocity and the band curvature to get the effective mass. The band curvature is computed as the second order perturbation in the kp method, for which Phil Hasnip and Peter Byrnes provided me with a piece a code to modify Castep.

There is a key word "momentum" that switch between one momentum interpretation or the other. It is under testing since both three approaches gives different answers.

ajm143 commented:
@pjh503 This is technically a CASTEP issue. But seems to fit better with this PR.

We seem now to have a ddome filename. I think that name could be confusing.
- ome are the optical matrix elements j,i
- dome is the diagonal optical matrix elements I=j.  Hence d stands for diagonal
- ddome second order k.p perturbation, I=j. Hence the first d stands for derivative, and the second d for diagonal. 

I'm worried that this might get messy if we use d to mean two different things. What are your thoughts?

pjh503 commented:
I agree that ddome_bin would be confusing, the question is what it should be called. I prefer "d" to mean "derivative", but unfortunately we've already used it for "diagonal". 

I would also prefer the name to reflect the fact that it is a derivative with respect to "dk", as we can imagine cases whether we look at the derivative with respect to other things (e.g. strain, phonon mode).

ajm143 commented:
.dome_dk ?  I agree we're using d to mean two different things here...  maybe we get away from omes and call it de_dk or similar?
.bands_dk ?

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+
+    epsilon_zero = 55.26349406_dp
+    ge = 10.24633_dp
+    kB=8.617333262E-5_dp
```
Put constants in the constant module please.

victorchanglee replied:
I moved all the constants to the constant module.

victorchanglee replied:
> How do we test your code? What version of CASTEP generates the necessary OptaDOS input files?

I will prepare some input/output examples for the code. There are two input files required to run some of the options in the code that are generated by a modified version of CASTEP 18.1

One is the free electron matrix, which I generated by modifying the spectral.f90 in CASTEP, in which the final state is substituted by a free electron planewave.

the second file is the band curvature which I also modified the spectral.f90 in CASTEP for this purpose. 

I think ideally both files should be written as a tool file in CASTEP but at the time, it was faster and easier for me to just modify the spectral.f90

ajm143 replied:
I think the CASTEP code freeze is coming up. When will your modifications to CASTEP be in a release ?

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+      width = (1.0_dp/11604.45_dp)*temp
+      qe_factor=1.0_dp/(4*pi**2*s_area)
+      norm_vac = gaussian(0.0_dp,width,0.0_dp)
+      kB=8.617333262E-5_dp
```
put contestants in constants module please.

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+    !Electric field and field emission
+  if(elec_field.gt.0.0_dp)then
+    call effect_wf
+else
```
correct tabbing please

victorchanglee replied:
fixed

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+
+    real(kind=dp) :: z,epsilon_zero
+
+    epsilon_zero = 55.26349406_dp
```
please put the constants in constants.f90

ajm143 commented on optados/examples/Photoemission/Cu.odi:
```diff
@@ -0,0 +1,20 @@
+s_area : 6.339744873
+slab_volume : 190.942338 
+elec_field : 0.001
+imfp_const : 19.0 
```
I think I can work out all of the other keywords, this one seems very mysterious. Is there a better name for it?

victorchanglee replied:
I changed s_area to surface_area, that should be more clear.

I will checked the name of the keywords to make them more clear.

ajm143 replied:
There's an argument that photoemission keywords relevant to that module /only/ might be prefixed by `PHOTOE_` or something like that?  Similar to CASTEP when you have `SPECTRAL_????` etc. I don't know what everyone feels about that -- but it's not a big deal to change up until we publish how to use the new module.

ajm143 commented on optados/examples/Photoemission/Cu.odi:
```diff
@@ -0,0 +1,20 @@
+TASK              : photoemission
+work_function : 4.556
+s_area : 6.339744873
+slab_volume : 190.942338 
```
Why isn't the slab volume calculated from the lattice vectors ?

ajm143 commented on optados/examples/Photoemission/Cu.odi:
```diff
@@ -0,0 +1,20 @@
+TASK              : photoemission
+work_function : 4.556
+s_area : 6.339744873
```
What does s_area mean? Slab area? If so, it would be better to call it slab_area to match slab_volume.

ajm143 commented on optados/examples/Photoemission/Cu.odi:
```diff
@@ -0,0 +1,20 @@
+momentum : crystal
+hybrid_linear : true
+temp : 300
+photo_output = be
```
What is be? Can we have a better name for this please?

victorchanglee replied:
I changed it to binding_energy. It is a keyword to generate a plot of the photoemission intensities as a function of the binding energy.

ajm143 commented on optados/examples/Si2_PDIS/Si2.pdis.dat:
```diff
@@ -0,0 +1,4013 @@
+ ##############################################################################
```
I don't think this should be checked in.

ajm143 commented on optados/src/algorithms.f90:
```diff
@@ -43,6 +43,8 @@ module od_algorithms
   public :: utility_cart_to_frac
   public :: utility_frac_to_cart
+  public :: utility_reciprocal_frac_to_cart
+  public :: utility_reciprocal_cart_to_frac
```
I don't understand these two functions. I think they're perfectly redundant.

`call utility_frac_to_cart(frac_rec,cart_rec,recip_lattice)`
returns the same as
`call subroutine utility_reciprocal_frac_to_cart(frac_rec,cart_rec,recip_lattice)`
no?

ajm143 commented on optados/src/electronic.f90:
```diff
@@ -437,6 +564,152 @@ subroutine elec_read_optical_mat
   end subroutine elec_read_optical_mat
 
+  !=========================================================================
+  subroutine elec_read_foptical_mat
```
shouldn't a free-electron matrix be called a .fem_bin file?
fome = free electron optical matrix elements? Is this what you mean?

victorchanglee replied:
yes, I initially just added the f to indicate that it is a file similar to the optical matrix element but with a free electron final state. I will changed the name of the file to make it more clear.

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+  !***************************************************************
+  subroutine make_optical_weights    
+    !***************************************************************
+    !This subroutine is copied from the OPTICS module
```
Why? Why not just use the one from the OPTICS module ? We really don't want the same function twice in the code.

victorchanglee replied:
Initially, the idea was to not touch the original OptaDOS code to avoid any issue. A lot of the optical functions that the photoemission module use are not public functions. Nevertheless, I will start making this functions publics and removing the duplicated ones from photo.f90

ajm143 replied:
I completely understand about not wanting to "break" the original OptaDOS code -- I had similar issues when I first coded for CASINO. However, it will make it easier going forward if we minimise the amount of duplicated code.  So thanks for that.

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+  !***************************************************************
+  subroutine intraband_photo
+    !***************************************************************
+    use od_electronic, only : optical_mat, elec_read_optical_mat, nbands,nspins, &
```
Is this a copy of an intraband function somewhere else? If so, could we use that ? Perhaps modify the original to be a function ?

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+  end subroutine intraband_photo
+  !***************************************************************
+  subroutine calc_epsilon_2_photo
+    !***************************************************************
```
Again, a copy of calc_epsilon_2? Why not just use that?

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+
+  !***************************************************************
+  subroutine calc_epsilon_1_photo
+    !***************************************************************
```
Copied function ?

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+
+  !***************************************************************
+  subroutine calc_refract_photo
+    !***************************************************************
```
Copy?

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+
+  !***************************************************************
+  subroutine calc_absorp_photo
+    !***************************************************************
```
Copy ?

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+  !***************************************************************
+    subroutine calc_reflect_photo
+  !***************************************************************
+  ! This subroutine calculates the reflection coefficient
```
Copy?

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+
+ !===============================================================================
+  subroutine setup_energy_scale
+ !===============================================================================
```
This is a copy? Why not call the original ?

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+  !===============================================================================
+  subroutine jdos_utils_calculate_delta(delta_temp)
+  !===============================================================================
+  ! Copy from the jdos_utils.f90 
```
Why not just call jdos_utlis ?

victorchanglee replied:
This function is slightly trickier to call.

In the original jdos_utils, the delta function is computed on the fly (it is not allocated) in a variable called dos_temp to generate the weighted_jdos which is integrated over the BZ.

In photoemission, I want to generate the photoemission intensities as function of k, so I need the Dirac delta function to be allocated.

I think the solution is to allocate the dos_temp in calculate_jdos.

ajm143 commented on optados/src/photo.f90:
```diff
@@ -0,0 +1,2722 @@
+  end subroutine transverse_energy_spread
+
+  subroutine binding_energy_spread
+  ! This subroutine applies a Gaussian broadenning to the binding energy
```
This looks like a copy of the function above? Why not make a function that does both ?

victorchanglee replied:
I made changes so now it only generates a plot of the photoemission intensities (QE) as a function of the binding energy, which is what experimentalist usually obtain from UPS/XPS/ARPES.

ajm143 reviewed (commented):
The looks good, and is almost totally in the OptaDOS coding style (despite us never publishing such).

There seems to me to be a lot of redundant code, in photo.f90 that could just call other functions instead. You shouldn't feel bad about generalising functions in other modules to make them do what we both want.   It will be much easier for us all to maintain.

How do we test your code? What version of CASTEP generates the necessary OptaDOS input files? 

I'm very keen to get this code into the stable branch.

victorchanglee committed:
Several improvements.

Several changes:

- Constants moved to constants.f90

-Merged the optical properties in the photoemission properties with the now public functions in optics.f90.

There are still a few functions that are copies with small changes.

I included all the input requires to run the photoemission module for a graphene sheet, including the free electron matrix.

I will try to update the code with further small changes.

ajm143 commented on optados/src/electronic.f90:
```diff
@@ -405,7 +405,7 @@ subroutine elec_read_band_curvature
 
-       ! Convert all band curvatures to eV Ang
+       ! Convert all band curvatures to eV Ang^2
        band_curvature=band_curvature*bohr2ang*bohr2ang*H2eV
```
Be very careful here. OptaDOS does a double conversion here bohr2ang twice because of an error in the units of the optical matrix elements coming out of CASTEP. Are you sure that the band curvature has the same error?

victorchanglee replied:
I think it has same error. I compared the calculated band curvature from a near-free electron band calculated as a chain of Li atoms in CASTEP with a polynomial fitting of the same band.

After unit conversion, the difference in curvature with both methods is around 0.02, which I think it is more likely to be an issue of the polynomial fitting.

jryates committed:
Merge branch 'develop' into develop
